### PR TITLE
(PE-14495) Remove pluginsync setting if upgrading to 1.4.x

### PR DESCRIPTION
**Pull request by MikaelSmith**

### CODE REVIEW

File: manifests/prepare/puppet_config.pp [hunk 1]
```diff
@@ -2,37 +2,58 @@
...
+        'config_version', 'manifest', 'modulepath',
+        # Settings that should be reset to defaults
+        'disable_warnings', 'vardir', 'rundir', 'libdir', 'confdir', 'ssldir', 'classfile']
+    } else {
+      $_removedSettings = []
+    }
+
+    # When upgrading to 1.4.x or later remove pluginsync
+    if ($package_version == undef and $old_packages) or (versioncmp($package_version, '1.4.0') >= 0) {
+      $removedSettings = $_removedSettings + ['pluginsync']
```
Comment (MikaelSmith) [inline manifests/prepare/puppet_config.pp]: I still want to guard this if the setting is already present.

**Comment by MikaelSmith** [comment]
Updated, ready to go.

**Comment by tphoney** [comment]
+1

**Comment by highb** [comment]
Ran this thorough a 2015.3.2->2015.3.x upgrade and I'm getting:
```
Info: Caching catalog for f4ogqvu5lus0moc.delivery.puppetlabs.net
      Info: Applying configuration version '1459980336'
      Error: /Stage[main]/Puppet_agent::Prepare::Puppet_config/Ini_setting[/pluginsync]: Could not evaluate: no implicit conversion of nil into String
      Error: /Stage[main]/Puppet_agent::Prepare::Puppet_config/Ini_setting[master/pluginsync]: Could not evaluate: no implicit conversion of nil into String
      Error: /Stage[main]/Puppet_agent::Prepare::Puppet_config/Ini_setting[agent/pluginsync]: Could not evaluate: no implicit conversion of nil into String
      Error: /Stage[main]/Puppet_agent::Prepare::Puppet_config/Ini_setting[main/pluginsync]: Could not evaluate: no implicit conversion of nil into String
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[puppetlabs-pepackages]: Dependency Ini_setting[/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[puppetlabs-pepackages]: Dependency Ini_setting[master/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[puppetlabs-pepackages]: Dependency Ini_setting[agent/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[puppetlabs-pepackages]: Dependency Ini_setting[main/pluginsync] has failures: true
      Warning: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[puppetlabs-pepackages]: Skipping because of failed dependencies
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki]: Dependency Ini_setting[/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki]: Dependency Ini_setting[master/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki]: Dependency Ini_setting[agent/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki]: Dependency Ini_setting[main/pluginsync] has failures: true
      Warning: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki]: Skipping because of failed dependencies
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki/rpm-gpg]: Dependency Ini_setting[/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki/rpm-gpg]: Dependency Ini_setting[master/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki/rpm-gpg]: Dependency Ini_setting[agent/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki/rpm-gpg]: Dependency Ini_setting[main/pluginsync] has failures: true
      Warning: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki/rpm-gpg]: Skipping because of failed dependencies
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki/rpm-gpg/RPM-GPG-KEY-puppetlabs]: Dependency Ini_setting[/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki/rpm-gpg/RPM-GPG-KEY-puppetlabs]: Dependency Ini_setting[master/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki/rpm-gpg/RPM-GPG-KEY-puppetlabs]: Dependency Ini_setting[agent/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki/rpm-gpg/RPM-GPG-KEY-puppetlabs]: Dependency Ini_setting[main/pluginsync] has failures: true
      Warning: /Stage[main]/Puppet_agent::Osfamily::Redhat/File[/etc/pki/rpm-gpg/RPM-GPG-KEY-puppetlabs]: Skipping because of failed dependencies
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Exec[import-RPM-GPG-KEY-puppetlabs]: Dependency Ini_setting[/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Exec[import-RPM-GPG-KEY-puppetlabs]: Dependency Ini_setting[master/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Exec[import-RPM-GPG-KEY-puppetlabs]: Dependency Ini_setting[agent/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Exec[import-RPM-GPG-KEY-puppetlabs]: Dependency Ini_setting[main/pluginsync] has failures: true
      Warning: /Stage[main]/Puppet_agent::Osfamily::Redhat/Exec[import-RPM-GPG-KEY-puppetlabs]: Skipping because of failed dependencies
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[pc_repo]: Dependency Ini_setting[/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[pc_repo]: Dependency Ini_setting[master/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[pc_repo]: Dependency Ini_setting[agent/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[pc_repo]: Dependency Ini_setting[main/pluginsync] has failures: true
      Warning: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[pc_repo]: Skipping because of failed dependencies
      Notice: /Stage[main]/Puppet_agent::Install/Package[puppet-agent]: Dependency Ini_setting[/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Install/Package[puppet-agent]: Dependency Ini_setting[master/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Install/Package[puppet-agent]: Dependency Ini_setting[agent/pluginsync] has failures: true
      Notice: /Stage[main]/Puppet_agent::Install/Package[puppet-agent]: Dependency Ini_setting[main/pluginsync] has failures: true
      Warning: /Stage[main]/Puppet_agent::Install/Package[puppet-agent]: Skipping because of failed dependencies
      Notice: Applied catalog in 0.74 seconds

    f4ogqvu5lus0moc.delivery.puppetlabs.net (centos5-64-agent) executed in 13.33 seconds
    Exited: 4
    Beaker::Host::CommandFailure: Host 'f4ogqvu5lus0moc.delivery.puppetlabs.net' exited with 4 running:
     puppet agent -t
    Last 10 lines of output were:
    	Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[pc_repo]: Dependency Ini_setting[master/pluginsync] has failures: true
    	Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[pc_repo]: Dependency Ini_setting[agent/pluginsync] has failures: true
    	Notice: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[pc_repo]: Dependency Ini_setting[main/pluginsync] has failures: true
    	Warning: /Stage[main]/Puppet_agent::Osfamily::Redhat/Yumrepo[pc_repo]: Skipping because of failed dependencies
    	Notice: /Stage[main]/Puppet_agent::Install/Package[puppet-agent]: Dependency Ini_setting[/pluginsync] has failures: true
    	Notice: /Stage[main]/Puppet_agent::Install/Package[puppet-agent]: Dependency Ini_setting[master/pluginsync] has failures: true
    	Notice: /Stage[main]/Puppet_agent::Install/Package[puppet-agent]: Dependency Ini_setting[agent/pluginsync] has failures: true
    	Notice: /Stage[main]/Puppet_agent::Install/Package[puppet-agent]: Dependency Ini_setting[main/pluginsync] has failures: true
    	Warning: /Stage[main]/Puppet_agent::Install/Package[puppet-agent]: Skipping because of failed dependencies
    	Notice: Applied catalog in 0.74 seconds
```

File: manifests/prepare/puppet_config.pp [hunk 2]
```diff
@@ -2,37 +2,59 @@
...
+
+    # When upgrading to 1.4.x or later remove pluginsync
+    if (($package_version == undef and $old_packages) or (versioncmp($package_version, '1.4.0') >= 0))
+        and !defined(Ini_setting["${section}/pluginsync"]) {
+      $removedSettings = $_removedSettings + ['pluginsync']
+    } else {
+      $removedSettings = $_removedSettings
+    }
+
+    if $removedSettings {
```
Comment (highb) [inline manifests/prepare/puppet_config.pp]: I can't remember, are arrays truthy in puppet now? Perhaps we need to be checking empty and not truth?

Comment (highb) [inline manifests/prepare/puppet_config.pp]: Ah, yep. Tested with Puppet 4:
```
# test.pp
$array = []
if $array {
  notify {'array is true':}
  }else{
  notify {'array is false':}
  }
```

```
~ puppet apply test.pp
WARN: Unresolved specs during Gem::Specification.reset:
      hiera (< 4, >= 2.0)
      json_pure (>= 0)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.
Notice: Compiled catalog for brandonhighmbp.corp.puppetlabs.net in environment production in 0.04 seconds
Notice: array is true
Notice: /Stage[main]/Main/Notify[array is true]/message: defined 'message' as 'array is true'
Notice: Applied catalog in 0.01 seconds
```

Comment (highb) [inline manifests/prepare/puppet_config.pp]: Although... the `.each` won't iterate over an empty array.

Comment (MikaelSmith) [inline manifests/prepare/puppet_config.pp]: I assumed incorrectly there.

File: manifests/prepare/puppet_config.pp [hunk 3]
```diff
@@ -2,37 +2,59 @@
...
+) {
   assert_private()
+  $old_packages = (versioncmp("${::clientversion}", '4.0.0') < 0)
 
-  $puppetconf = $::puppet_agent::params::config
-  file { $puppetconf:
-    ensure => file,
-    source => $::puppet_config,
+  if $old_packages {
+    $puppetconf = $::puppet_agent::params::config
```
Comment (MikaelSmith) [inline manifests/prepare/puppet_config.pp]: Oops, this should always be set.

**Comment by MikaelSmith** [comment]
Ok, made updates.

File: manifests/prepare/puppet_config.pp (hunk 2 above)
Comment (highb) [inline manifests/prepare/puppet_config.pp]: Yeah, we can probably remove the if. As for my comment about the `.each`, I'm just trying to figure out where `no implicit conversion of nil into String` is coming from.

**Comment by highb** [comment]
@MikaelSmith The puppet.conf dep is still causing issues:
```
  1) puppet_agent supported operating systems on centos-5-x86_64 valid package_versions should compile into a catalogue without dependency cycles
     Failure/Error: it { is_expected.to compile.with_all_deps }
       error during compilation: Could not retrieve dependency 'File[/etc/puppetlabs/puppet/puppet.conf]' of Ini_setting[/pluginsync]
```

**Comment by MikaelSmith** [comment]
Oh, that's interesting.

**Comment by MikaelSmith** [comment]
Apparently the file needs to be managed. Hopefully fixed by ensuring it's a file if not already managed.

**Comment by highb** [comment]
:+1: agent_upgrade test passed with this change.

**Pull request merged**